internal/kafka/internal/services/kafka.go

```diff
@@ -776,6 +787,59 @@ func (k *kafkaService) GetManagedKafkaByClusterID(clusterID string) ([]managedka
 	return res, nil
 }
 
+func (k *kafkaService) GenerateReservedManagedKafkasByClusterID(clusterID string) ([]managedkafka.ManagedKafka, *errors.ServiceError) {
+	reservedKafkas := []managedkafka.ManagedKafka{}
+
+	if !k.dataplaneClusterConfig.IsDataPlaneAutoScalingEnabled() {
+		return reservedKafkas, nil
+	}
+
+	cluster, svcErr := k.clusterService.FindClusterByID(clusterID)
+	if svcErr != nil {
+		return nil, svcErr
+	}
+	if cluster == nil {
+		return nil, apiErrors.GeneralError("failed to generate reserved managed kafkas for clusterID %s: clusterID not found", clusterID)
+	}
+	if cluster.Status != api.ClusterReady {
+		logger.Logger.Warningf("ClusterID '%s' is not ready. Its status is '%s'. Returning an empty list of reserved managed kafkas", clusterID, cluster.Status)
+		return reservedKafkas, nil
+	}
+
+	strimziVersions, err := cluster.GetAvailableAndReadyStrimziVersions()
+	if err != nil {
+		return nil, errors.NewWithCause(apiErrors.ErrorGeneral, err, "failed to generate reserved managed kafkas for clusterID %s: error finding ready strimzi versions", clusterID)
+	}
+	if len(strimziVersions) == 0 {
+		return nil, apiErrors.GeneralError("failed to generate reserved managed kafkas for clusterID %s: no ready strimzi versions found")
+	}
+	latestStrimziVersion := strimziVersions[len(strimziVersions)-1]
```

Comment:
We could have a `GetLatestStrimziVersion` as well which we could use here as:
```go
latestStrimziVersion, strimziVersionErr := cluster.GetLatestStrimziVersion()
```

Reply:
In any case it would be a `

> should we ignore querying the database for the kafkas whose id begins with `reserved-kafka-` and only log the status info of the rejected kafkas? This newly added log could later on be replaced by a metric to see how many reserved streaming units are there and are in a `ready` or `rejected` status.


> should we ignore querying the database for the kafkas whose id begins with `reserved-kafka-` and only log the status info of the rejected kafkas? This newly added log could later on be replaced by a metric to see how many reserved streaming units are there and are in a `ready` or `rejected` status.

That is a good question. This method also persists information of the kafka in the DB.
With the current implementation an error will be logged and the iteration continue:
```
		if getErr != nil {
			glog.Error(errors.Wrapf(getErr, "failed to get kafka cluster by id %s", ks.KafkaClusterId))
			continue
		}
```

In any case if we were to do any comparison the ID comparison seems a little bit fragile. If we were to improve that maybe it would be a good idea for the data plane to send us back if it is a reserved kafka or not? (as we do when we provide that info to them)

Reply:
In this case it would be a `GetLatestAvailableAndReadystrimziVersion` method.
I will implement `GetLatestAvailableAndReadystrimziVersion` and `GetLatestAvailableReadystrimziVersion` methods.

Reply:
The methods have been added